test/cpp/jit/test_concat_opt.cpp

```diff
@@ -0,0 +1,451 @@
+      ->check_count("= aten::cat(", 0, /*exactly*/ true)
+      ->check_count("= aten::empty(", 0, /*exactly*/ true)
+      ->run(*graph);
+}
+
+TEST(OptimizeConcatTest, ConcatWithoutResultShape) {
+  auto graph = std::make_shared<Graph>();
+
+  const std::string input =
+      R"IR(
+        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu),
+              %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)):
+          %2 : int = prim::Constant[value=0]()
+          %3 : float = prim::Constant[value=0.5]()
+          # CHECK: clamp_max
+          %4 : Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::clamp_max(%0, %3)
+          # CHECK: clamp_max
+          %5 : Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::clamp_max(%1, %3)
+          # CHECK: prim::ListConstruct
+          %6 : Tensor[] = prim::ListConstruct(%4, %5)
+          # CHECK: aten::cat
+          %7 : Tensor = aten::cat(%6, %2)
+          return (%7)
+      )IR";
+  parseIR(input, graph.get());
+  std::vector<at::Tensor> inputs = {
+      at::rand({64, 56, 56}, at::kCPU), at::rand({32, 56, 56}, at::kCPU)};
+  auto orig_outputs = runGraph(graph, inputs);
+
+  ExpandConcatAndEliminateRedundancy(graph);
+  graph->lint();
+  auto opt_outputs = runGraph(graph, inputs);
+
+  checkOutputs(orig_outputs, opt_outputs);
+
+  // No optimizations should have happened in this case since the output
+  // shape of `aten::cat` is not known.
+  testing::FileCheck().run(input, *graph);
+}
+
+TEST(OptimizeConcatTest, ConcatWithoutInputShape) {
+  auto graph = std::make_shared<Graph>();
+
+  const std::string input =
+      R"IR(
+        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu),
+              %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)):
+          %2 : int = prim::Constant[value=0]()
+          %3 : float = prim::Constant[value=0.5]()
+          # CHECK: clamp_max
+          %4 : Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::clamp_max(%0, %3)
+          # CHECK: clamp_max
+          %5 : Tensor = aten::clamp_max(%1, %3)
+          # CHECK: prim::ListConstruct
+          %6 : Tensor[] = prim::ListConstruct(%4, %5)
+          # CHECK: aten::cat
+          %7 : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%6, %2)
+          return (%7)
+      )IR";
+  parseIR(input, graph.get());
+  std::vector<at::Tensor> inputs = {
+      at::rand({64, 56, 56}, at::kCPU), at::rand({32, 56, 56}, at::kCPU)};
+  auto orig_outputs = runGraph(graph, inputs);
+
+  ExpandConcatAndEliminateRedundancy(graph);
+  graph->lint();
+  auto opt_outputs = runGraph(graph, inputs);
+
+  checkOutputs(orig_outputs, opt_outputs);
+
+  // No optimizations should have happened in this case since the output
+  // shape of `aten::cat` is not known.
```

Comment:
output shape of %7 is known:

```
%7 : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%6, %2)
```
Or did I miss something?

Reply:
Oh, the comment is incorrect here. Here it does not get optimized because one of the inputs of `cat`, `%5`, does not have shape info. 

Sorry about that. Will fix the comment.

Reply:
Fixed comment.